### PR TITLE
Add persisted query cache

### PR DESCRIPTION
**Pull request by eliias**
## Todos 
- [ ] Tests
- [ ] Documentation
- [ ] Typings
- [ ] Changelog

### CODE REVIEW

**Comment by eliias** [comment]
> @eliias as a super minor thing, i would just not commit the `dist` folder -- we need to build it on `master` anyways (to ensure it's up to date) and there's an rxdb lint rule to prevent merging in changes to `dist` via pr

👍  hehe… I already got confused by all the compiled files

File: src/plugins/utils/utils-object.ts
Comment (TristanH) [inline src/plugins/utils/utils-object.ts:34]: lol, those damn linters eh

File: src/rx-query.ts
Comment (TristanH) [inline src/rx-query.ts:35]: 👌

File: src/rx-query.ts
Comment (mitchelljustin) [inline src/rx-query.ts:333]: lol nice, dunno why they chained promises in the original code

**Comment by TristanH** [comment]
@mitchelljustin pls keep in mind we're punting on implementing limit buffer with persistence right now. this means you lose the limit buffer across sessions for now. i still hope to get it done this week, but it can't block this PR -- it's too important!

File: src/rx-query.ts
Comment (mitchelljustin) [inline src/rx-query.ts:180]:
```suggestion
    public _persistentQueryCacheResultLwt?: string = undefined; // lwt = latest write time
```

File: src/rx-query.ts
Comment (mitchelljustin) [inline src/rx-query.ts:534]: what's a dexie

File: src/rx-query.ts
Comment (TristanH) [inline src/rx-query.ts:207]: what is the purpose of this new field? I guess `Object.keys(docsMap)` isn't ordered? why not just `docsData.map((d)=>d.primary)`?

File: src/rx-query.ts
Comment (TristanH) [reply src/rx-query.ts:333]: i don't understand this change.. is it just rewriting with await instead of then?

Comment (TristanH) [reply src/rx-query.ts:333]: maybe there's a reason he didn't want to use an `async` function?

Comment (eliias) [reply src/rx-query.ts:333]: My IDE "simplified" that for me.

File: src/rx-query.ts
Comment (mitchelljustin) [inline src/rx-query.ts:523]: why are you setting _persistentQueryCacheResult  twice?

File: src/rx-query.ts
Comment (mitchelljustin) [inline src/rx-query.ts:531]: is this a pretty fast and well indexed query? i'm assuming it is since you're gonna be running it for each query using pers. cache

File: src/rx-query.ts
Comment (eliias) [reply src/rx-query.ts:534]: One of the supported storage engines, he uses `dexie` in his tests (and Loki).

File: src/rx-query.ts
Comment (eliias) [reply src/rx-query.ts:207]: Not strictly necessary, just a minor optimization. We only need the keys, and there is no reason to iterate over the result set later.

File: src/rx-query.ts
Comment (eliias) [reply src/rx-query.ts:523]: No reason, oversight.

File: src/rx-query.ts
Comment (eliias) [reply src/rx-query.ts:531]: `[lwt+id]` is a built-in default index on every storage.

File: src/rx-query.ts
Comment (mitchelljustin) [inline src/rx-query.ts:563]: it's a bit strange to modify some captured variable `docsData` here inside of a `filter()` closure. normally those are separate

File: src/rx-query.ts
Comment (mitchelljustin) [inline src/rx-query.ts:585]: very cool

File: src/rx-query.ts
Comment (eliias) [reply src/rx-query.ts:531]: This is doing the same, I am good with either.

```ts
const nonRestoredDocIds: string[] = [];
for (const docId of persistedQueryCacheIds) {
  // first try to fill from docCache
  const docData = this.collection._docCache.getLatestDocumentDataIfExists(docId);
  if (docData && this.doesDocumentDataMatch(docData)) {
    docsData.push(docData);
  }

  if (!docData) {
    nonRestoredDocIds.push(docId);
  }
}
```

File: src/rx-query.ts
Comment (mitchelljustin) [inline src/rx-query.ts:182]:
```suggestion
    public _persistentQueryCacheBackend?: QueryCacheBackend;
```

File: src/rx-query.ts
Comment (eliias) [reply src/rx-query.ts:563]: This is doing the same, I am good with either.

```ts
const nonRestoredDocIds: string[] = [];
for (const docId of persistedQueryCacheIds) {
  // first try to fill from docCache
  const docData = this.collection._docCache.getLatestDocumentDataIfExists(docId);
  if (docData && this.doesDocumentDataMatch(docData)) {
    docsData.push(docData);
  }

  if (!docData) {
    nonRestoredDocIds.push(docId);
  }
}
```

File: src/rx-query.ts
Comment (mitchelljustin) [reply src/rx-query.ts:182]: to disambiguate from memory query cache

File: src/rx-query.ts
Comment (mitchelljustin) [reply src/rx-query.ts:35]: 👌🏻

File: src/plugins/utils/utils-object.ts
Comment (mitchelljustin) [reply src/plugins/utils/utils-object.ts:34]: new version of Jetbrains forces a plugin on you that 'corrects' spelling. i dislike it, it's often wrong or pedantic. maybe now i know how it feels 😂

File: src/rx-query.ts
Comment (TristanH) [inline src/rx-query.ts:519]: can you just return early here if value is null, or am i missing something?

File: src/rx-query.ts
Comment (mitchelljustin) [inline src/rx-query.ts:37]: do you even need the T type param here? isn't it always just a cache for `string | string[]`?

Comment (eliias) [reply src/rx-query.ts:37]: IIRC I did this to be compatible with the `localforage` interface. There are probably nicer ways to write this.

File: src/rx-query.ts
Comment (TristanH) [inline src/rx-query.ts:535]: i miiight suggest doing like `lwt - 100` to avoid any potential races, but i don't feel strongly

Comment (eliias) [reply src/rx-query.ts:535]: Yeah, I don't know if there are any guarantees that the same query can't be executed concurrently. Maybe we should just do this. Atm, I take the value from `const lwt = rxQuery._result?.time ?? 0;`, which is just the `lwt` of the query result we persist.

File: src/rx-query.ts
Comment (mitchelljustin) [reply src/rx-query.ts:207]: but Object.keys() is super fast, no?

File: src/rx-query.ts
Comment (mitchelljustin) [inline src/rx-query.ts:260]: so what happens now with counts? are they permanently wrong until the persistent query cache is busted?

File: src/rx-query.ts
Comment (TristanH) [inline src/rx-query.ts:531]: i am NOT saying we do this now, but it does seem (in theory) fairly wasteful that this LWT will run once for every cached query.

i wonder if you could have something like the docCache that caches all documents `lwt>X`. then if another query is trying to fetch `lwt>Y` (Y>X) you can just loop through lwt>X. if another query tries to fetch `lwt>Z` (X>Z), you filter the db for `X>lwt>Z` only, and get the rest from `lwt>X` :P

File: src/rx-query.ts
Comment (eliias) [reply src/rx-query.ts:260]: I currently persist the result of a `count` query every time there is a change handled by event-reduce (and not actually querying). I couldn't produce a mistake, but there is no correction mechanism either. Feels scary, have no solution to offer (except storing all ID's of the count query and doing the same we do with regular queries, but that sounds like adding lots of slowness).

File: src/rx-query.ts
Comment (mitchelljustin) [reply src/rx-query.ts:333]: AFAIK there's no functional difference between chaining promises and async/await in JS. IMO using async/await reads and codes much nicer

File: src/rx-query.ts
Comment (mitchelljustin) [reply src/rx-query.ts:531]: excellent

File: src/rx-query.ts
Comment (mitchelljustin) [reply src/rx-query.ts:563]: that definitely reads nicer to me if i ever had to understand & change this code.

File: src/rx-query.ts
Comment (eliias) [reply src/rx-query.ts:531]: Hm, I think you still need to run `lwt > last_of(set_X)` to also fetch updates since the first query ran. But in general I agree, we could do this per collection instead of per query.

File: src/rx-query.ts
Comment (mitchelljustin) [inline src/rx-query.ts:543]: yeah IDK what's up with the original code, this is a pretty basic Programming 101 mistake

File: src/rx-query.ts
Comment (mitchelljustin) [inline src/rx-query.ts:743]: we have to await here? cannot just update pers query cache async?

Comment (eliias) [reply src/rx-query.ts:746]: This is for the sake of consistency. We wait till the result of `count` is stored in `Cache`, before we return the result to user code. Otherwise, we might show a value that we never persist, and the next query with the same ID could increment from a different value in case of any change events.

**Review by mitchelljustin — COMMENTED** [review]
test comment

**Review by mitchelljustin — COMMENTED** [review]
test comment

File: src/rx-query.ts
Comment (mitchelljustin) [reply src/rx-query.ts:746]: OK

File: src/rx-query.ts
Comment (mitchelljustin) [reply src/rx-query.ts:260]: mmm, so error(persistent count) == error(event-reduce). i can live with that, event-reduce does make mistakes but they tend to be rare/edge-casey.

File: src/rx-query.ts
Comment (mitchelljustin) [reply src/rx-query.ts:37]: OK

File: src/rx-query.ts
Comment (TristanH) [reply src/rx-query.ts:260]: event-reduce shouldn't make mistakes after we upgrade to his newest version:

https://github.com/pubkey/event-reduce/issues/444#issuecomment-1829793615

(which i'm doing today)

File: src/rx-query.ts
Comment (TristanH) [reply src/rx-query.ts:333]: my bias is to not change stuff if there's no need. this is a fork we have to keep up to date, after all

File: src/rx-query.ts
Comment (eliias) [reply src/rx-query.ts:207]: It is still `O(n)` and not necessary.

File: src/rx-query.ts
Comment (TristanH) [inline src/rx-query.ts:539]: i wonder if there's some number of changedDocs that we just want to give up and re-exec the query. i'm imagining a user bulk updating thousands of docs....

File: src/rx-query.ts
Comment (TristanH) [reply src/rx-query.ts:746]: i wonder if we should just skip persistence for counts<X. i guess those counts could still be slow 🤔  i guess we can just choose where to persist vs not. for example, i might suggest we only persist counts of non-built-in queries

File: test/unit/rx-query.test.ts
Comment (mitchelljustin) [inline test/unit/rx-query.test.ts:1633]: 🙌🙌🙌

File: test/unit/rx-query.test.ts
Comment (mitchelljustin) [inline test/unit/rx-query.test.ts:1704]: huge

File: src/rx-query.ts
Comment (eliias) [reply src/rx-query.ts:539]: You would still need to explicitly enable the cache for such queries, but it is not too hard to change if we want it.

File: test/unit/rx-query.test.ts
Comment (mitchelljustin) [reply test/unit/rx-query.test.ts:1607]: in particular i'd like to test whether the logic that writes to pers q cache works with the logic that reads from it

File: test/unit/rx-query.test.ts
Comment (mitchelljustin) [inline test/unit/rx-query.test.ts:1815]: can't you put these in a beforeAfter()?

Comment (mitchelljustin) [reply test/unit/rx-query.test.ts:1815]: maybe kind of annoying to do with a global variable. idc that much

File: src/rx-query.ts
Comment (mitchelljustin) [reply src/rx-query.ts:532]: it's also a bit strange that the memory query cache is a property of the RX collection, but the persistent query cache is a property of each individual query. might could kill two birds with one stone if you get around to it (but not blocking this PR)

**Review by mitchelljustin — APPROVED** [review]
incredible work, some comments but from my side nothing is blocking shipping. probably best to get tristan's approval too, though

File: src/rx-query.ts
Comment (TristanH) [inline src/rx-query.ts:826]: ooc what's the rationale behind this?

File: src/rx-query.ts
Comment (TristanH) [inline src/rx-query.ts:917]: likewise i know his formatting here sucks, but for the sake of future upgrades to rxdb15, 16, 17, etc can we keep the formatting the same?

File: src/rx-query.ts
Comment (TristanH) [inline src/rx-query.ts:762]: does this actually need to block `__ensureEqual`? i think it's safe to not...

File: src/rx-query.ts
Comment (eliias) [reply src/rx-query.ts:826]: If the query cache was persisted more than an hour ago, and either event-reduce or exec of the query runs, it persists the current query cache. This eventually refreshes the cache, but at a very slow pace. I don't know what a good TTL is for this, I just picked an arbitrary value. Could be a day as well, or 1min.

File: test/unit/rx-query.test.ts
Comment (eliias) [reply test/unit/rx-query.test.ts:1815]: Would be nice. I don't even know which test library RxDB uses. I just copy & pasted code from other tests.

File: test/unit/rx-query.test.ts
Comment (eliias) [reply test/unit/rx-query.test.ts:1728]: Hm… I think we should use a comment, I am re-using a function that he wrote (https://github.com/readwiseio/rxdb_fork/blob/7607a8b9b2346f5de962eac9d2e04045a36d599c/test/unit/cache-replacement-policy.test.ts#L19). I just extracted it into a helper file, so that I can use it.

File: src/rx-query.ts
Comment (eliias) [reply src/rx-query.ts:532]: I think we can do it, but I believe it is not super simple. For example, we need to block (enqueue) and dedup the retrieval across all queries that query the same collection. When a query with an older timestamp starts after a query with a newer timestamp, we need to cancel the request and initiate a new one.

File: src/rx-query.ts
Comment (eliias) [reply src/rx-query.ts:917]: yeah… i don't mind reverting this to its previous state, merging upstream changes is going to be fun anyway

File: src/rx-query.ts
Comment (eliias) [reply src/rx-query.ts:762]: It doesn't, I just defaulted to the safest thing. FWIW I had no issues with writes blocking UI.

File: src/rx-query.ts
Comment (TristanH) [inline src/rx-query.ts:587]: ugh @eliias i think that this approach actually has the same fundamental flaw the event-reduce library has with limit queries -- which we solve with the limit buffer.

imagine a world where you have 10000 docs match, but LIMIT 10. then those first 10 docs are deleted/changed (or even one of them) so that they no longer match the query. the correct thing to do here would be to load the next 10 docs that match the query. but you never cached those other 9990 that weren't in the limit. and those 9990 didn't change. so how do you get them?

I think you inevitably have to re-execute the whole query in this case. Otherwise your results will be incorrect. I think an easy way to check is just to make sure that your limit is equal to the length of your docsData. If it is, you just don't use the persisted query cache.

(By the way, this case isn't super common.)

(Also btw, persisting the limit buffer should let us mostly fix this...)

Comment (TristanH) [reply src/rx-query.ts:587]: Sorry I didn't think of this earlier :/

File: src/rx-query.ts
Comment (eliias) [reply src/rx-query.ts:762]: Ah… I correct myself here, I actually had a pretty good reason to await at this point, it just doesn't make sense when persistence is done conditionally.

File: test/unit/rx-query.test.ts
Comment (eliias) [reply test/unit/rx-query.test.ts:1607]: Busting the in-memory query-cache and re-executing the query is not going to work (aka not supported by this implementation). There are similar test cases that somewhat do what you suggest here, but they are focusing on new instances of `RxQuery` that have the same `persistentQueryId` (which is just a hashed query string). In such cases, the query result is actually restored.

1. The persistent cache is not restored lazily, but only at query creation time. It then fills up the in-memory cache. This is fine, as we create and execute queries around the same time.
2. The other reason I did not implement this has to do with the life cycle of the in-memory query cache. It is never busted completely, but instead a query only gets evicted from the cache when the cleanup marks it for removal. In such cases we could theoretically benefit from restoring again, but it is preferable to make sure the query never gets evicted in the first place. Tristan and I had a brief conversation about modifying our cache policy to guarantee that.

File: src/rx-query.ts
Comment (TristanH) [reply src/rx-query.ts:620]: copying my quick thoughts from slack here for visibility:

> I have instrumented the cost of a “write”, and it takes about 40-50ms. It is quite slow in comparison to when we’d write directly to IndexedDB (this is localforage), but it is not something a user would notice.

This is my main concern. You need to profile this on non-m1/m2 macbooks, as well as iphone and (especially) android devices

> Based on our other conversations on this PR I expect this to be a somewhat controversial discussion (sync writes), but I prefer this. Storing the limit buffer itself is just moving the problem elsewhere, because we then need to persist the limit buffer every time the result set changes as well.
This is because a user could also indirectly move an item that is in the limit buffer (aka the next skip: 50 query) from inbox -> later. When we then restore a stale limit buffer (the same way I restore the stale query cache in this PR) it would be returning wrong results as well.

limit buffer has a mechansim to handle this already. it simply re-execs the full query if something in the buffer has changed. it’s not perfect, but i think in practice in our app 95%+ of the time stuff won’t re-exec. so yes this is true:

> If we instead bust the query cache every time we discover a mismatch (different counts), we essentially trade write latency with query latency. I think this on its own would be fine, especially because we mostly restore the query cache to mitigate the cold-start issue before the memory storage initialization is done. Without continuously persisting the query cache, we loose this benefit on the next reload after a change. So the worst case here is that a user has a slow startup time on every other tab/window.

but you’re trading N writes for N queries every time a relevant doc updates, for a re-exec 5% (or less) of the time. Who is this for?

File: src/rx-query.ts
Comment (TristanH) [reply src/rx-query.ts:232]: i will tie break here and say lets just keep docKeys

File: src/rx-query.ts
Comment (eliias) [inline src/rx-query.ts:806]: 👨‍🍳 💋

**Pull request merged**
